staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldmanager
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal"
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+)
+
+var (
+	scaleGroupVersion = schema.GroupVersion{Group: "autoscaling", Version: "v1"}
+	scalePath         = fieldpath.MakePathOrDie("spec", "replicas")
+)
+
+// SubresourceScaler manages the conversion of managed fields between a main
+// resource and subresources
+type SubresourceScaler struct {
+	parentEntries      []metav1.ManagedFieldsEntry
+	parentGroupVersion schema.GroupVersion
+	parentPath         fieldpath.Path
+}
+
+// ScalePathMappings maps a group/version to the replicas path in a scale subresource
+type ScalePathMappings map[schema.GroupVersion]fieldpath.Path
+
+// NewSubresourceScaler creates a new SubresourceScaler
+func NewSubresourceScaler(parentEntries []metav1.ManagedFieldsEntry, parentGroupVersion schema.GroupVersion, parentPath fieldpath.Path) (*SubresourceScaler, error) {
+	return &SubresourceScaler{
+		parentEntries:      parentEntries,
+		parentGroupVersion: parentGroupVersion,
+		parentPath:         parentPath,
+	}, nil
+}
+
+// ToSubresource filter the managed fields of the main resource and convert
+// them so that they can be handled by a subresource.
+// For the managed fields that have a path that could be handled by a
+// subresource this method performs two changes:
+// 1. The APIVersion is changed to the APIVersion of the subresource
+// 2. The Fields are transformed to the path of that subresource
+func (h *SubresourceScaler) ToSubresource() ([]metav1.ManagedFieldsEntry, error) {
+	managed, err := internal.DecodeObjectManagedFields(h.parentEntries)
+	if err != nil {
+		return nil, err
+	}
+
+	f := fieldpath.ManagedFields{}
+	t := map[string]*metav1.Time{}
+	for manager, versionedSet := range managed.Fields() {
+
+		if versionedSet.Set().Has(h.parentPath) {
```

Comment:
`parentPath` depends on the version of the object (different versions can have the replicas path in different location), and you're not looking at the version of the deployment here.


This needs to look for the path in `replicasPathInDeployment` mapping for the version of that `versionedSet`.